src/amberscript/buffer.h

```diff
@@ -0,0 +1,70 @@
+// Copyright 2018 The Amber Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SRC_AMBERSCRIPT_BUFFER_H_
+#define SRC_AMBERSCRIPT_BUFFER_H_
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "src/datum_type.h"
+#include "src/value.h"
+
+namespace amber {
+namespace amberscript {
+
+enum class BufferType : uint8_t {
+  kColor = 0,
+  kDepth,
+  kFramebuffer,
+  kIndex,
+  kSampled,
+  kStorage,
+  kUniform,
+  kVertex
```

Comment:
I think we should remove `kFramebuffer` and `kSampled` and add `kStorageTexel`, `kUniformTexel`.
If I understand this correctly, it should include
- color attachment and depth/stencil attachment for framebuffer
- vertex, index
- [descriptors](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#descriptorsets-types)

Reply:
And I want to doublecheck, do `kColor` and `kDepth` mean color attachment and depth/stencil attachment for framebuffer that we will probe?

Reply:
Why would descriptors be here? They aren't a buffer are they? I'm trying to keep this agnostic from Vulkan, so want to avoid things like attachments.

kFramebuffer is how I know the buffer is a framebuffer and can be used in the subsequent FRAMEBUFFER calls inside a PIPELINE. Does a framebuffer need to be defined by multiple buffers in Vulkan?

Reply:
Storage and SampledImage are descriptors.
Framebuffer can have color and depth/stencil buffers.

I will file a bug to discuss how we should change those types.